README.md
# ntlm-client

A node.js NTLM client with support for NTLM and NTLMv2 authentication

```
npm install ntlm-client
```

## Example Usage
```javascript
const ntlmRequest = require('@elasticio/ntlm-client').request;

(async function makeExampleRequest() {
  const { response } = await ntlmRequest({
    username: 'MYDOMAIN\SomeUser',
    password: 'P@$$word1!',
    uri: 'https://api.someservice.com',
    method: 'POST',
    request: {
      json: true,   // Example of parameter passed to request.js
      body: {
        foo: 'bar'
      },
      headers: {
        Date: 'Today'
      }
    }
  });
  console.log(`Received status code: ${response.statusCode} Body: ${response.body}`);
})()
```

## API

#### request(options)

A convenience function that tries to authenticate against a given URL using the `request` module.
If NTLM is not supported, it will fallback to Basic Auth.

* Arguments
  * `options` an object holding below options for the authentication process:
    * `uri` the target URL
    * `method` the HTTP verb
    * `username`
    * `password`
    * `request` this is optional. An object that holds options that should be passed to the request instance
* Returns
  * `Promise` when resolved, the `request` instance, the result and the response body will be passed

---------------------------------------

#### createType1Message([workstation, target])

Creates a type 1 NTLM message to initialize the NTLM handshake

* Arguments
  * `workstation` Optional. If `undefined`, `os.hostname()` will be used
  * `target` Optional. This is the domain/host we are trying to authenticate against.
* Returns
  * `string` Complete NTLM string that should be sent to the server in the `Authentication` header

---------------------------------------

#### decodeType2Message(str)

Decodes a type 2 message received from the server including the NTLM challenge

* Arguments
  * `str` Either the base64 encoded type 2 message, or the complete `WWW-Authenticate` header, or an object containg the response headers (`http.IncomingMessage`)
* Returns
  * `type2Message` An object containing the following information about the received type 2 message: `flags`, `encoding`, `version`, `challenge`, `targetName`, `targetInfo`.

---------------------------------------

#### createType3Message(type2Message, username, password[, workstation, target])

Creates a type 3 message based on the type 2 message received from the server.

* Arguments
  * `type2Message` The decoded type 2 message object
  * `username`
  * `password`
  * `workstation` Optional. If falsy, `os.hostname()` will be used
  * `target` Optional. If falsy, the target name from the type 2 message will be used. This is the domain/host we are trying to authenticate against.
* Returns
  * `string` Complete NTLM string that should be sent to the server in the `Authentication` header


CHANGELOG.md
## 1.0.0 (March 5, 2020)
* Elastic.io GmbH re-release
* Add integration tests
* Update dependencies
* Don't fail if request object is omitted.
* Add support for usernames of the form 'SomeDomain\SomeUser'.
* Don't mask errors
* Don't supply an `Accept` header


spec-integration/integration-tests.spec.js
const fs = require('fs');
const { expect } = require('chai');

const ntlmRequest = require('../lib/index').request;

if (fs.existsSync('.env')) {
    // eslint-disable-next-line global-require
    require('dotenv').config();
}

describe('Integration Tests', function() {
    let username;
    let password;
    let url;

   this.timeout(50000);

   ['NTLM_V2', 'NTLM_V1', 'BASIC'].forEach(authType => {
       if(process.env[`${authType}_USERNAME`] && process.env[`${authType}_PASSWORD`] && process.env[`${authType}_URL`]) {
           console.log(`Detected Authorization Configuration For ${authType}`);
           describe(`${authType} Tests`, function () {
               before(() => {
                   username = process.env[`${authType}_USERNAME`];
                   password = process.env[`${authType}_PASSWORD`];
                   url = process.env[`${authType}_URL`];
               });

               it('Correct Password', async() => {
                   const { response } = await ntlmRequest({
                       username,
                       password,
                       uri: url,
                       method: 'GET'
                   });

                   expect(response.statusCode).to.be.equal(200);
               });

               it('Incorrect Password', async() => {
                   const { response } = await ntlmRequest({
                       username,
                       password: 'Wrong Password',
                       uri: url,
                       method: 'GET'
                   });

                   expect(response.statusCode).to.be.equal(401);
               });
           });
       } else {
           console.log(`Missing Required Authorization Configuration For ${authType}`);
       }
   });
});


lib/hash.js
'use strict';

const crypto = require('crypto');

function createLMResponse(challenge, lmhash) {
	let buf = new Buffer(24),
		pwBuffer = new Buffer(21).fill(0);

	lmhash.copy(pwBuffer);

	calculateDES(pwBuffer.slice(0, 7), challenge).copy(buf);
	calculateDES(pwBuffer.slice(7, 14), challenge).copy(buf, 8);
	calculateDES(pwBuffer.slice(14), challenge).copy(buf, 16);

	return buf;
}

function createLMHash(password) {
	let buf = new Buffer(16),
		pwBuffer = new Buffer(14),
		magicKey = new Buffer('KGS!@#$%', 'ascii');

	if (password.length > 14) {
		buf.fill(0);
		return buf;
	}

	pwBuffer.fill(0);
	pwBuffer.write(password.toUpperCase(), 0, 'ascii');

	return Buffer.concat([
			calculateDES(pwBuffer.slice(0, 7), magicKey),
			calculateDES(pwBuffer.slice(7), magicKey)
	]);
}

function calculateDES(key, message) {
	let desKey = new Buffer(8);

	desKey[0] = key[0] & 0xFE;
	desKey[1] = ((key[0] << 7) & 0xFF) | (key[1] >> 1);
	desKey[2] = ((key[1] << 6) & 0xFF) | (key[2] >> 2);
	desKey[3] = ((key[2] << 5) & 0xFF) | (key[3] >> 3);
	desKey[4] = ((key[3] << 4) & 0xFF) | (key[4] >> 4);
	desKey[5] = ((key[4] << 3) & 0xFF) | (key[5] >> 5);
	desKey[6] = ((key[5] << 2) & 0xFF) | (key[6] >> 6);
	desKey[7] = (key[6] << 1) & 0xFF;

	for (let i = 0; i < 8; i++) {
		let parity = 0;

		for (let j = 1; j < 8; j++) {
			parity += (desKey[i] >> j) % 2;
		}

		desKey[i] |= (parity % 2) === 0 ? 1 : 0;
	}

	let des = crypto.createCipheriv('DES-ECB', desKey, '');
	return des.update(message);
}

function createNTLMResponse(challenge, ntlmhash) {
	let buf = new Buffer(24),
		ntlmBuffer = new Buffer(21).fill(0);

	ntlmhash.copy(ntlmBuffer);

	calculateDES(ntlmBuffer.slice(0, 7), challenge).copy(buf);
	calculateDES(ntlmBuffer.slice(7, 14), challenge).copy(buf, 8);
	calculateDES(ntlmBuffer.slice(14), challenge).copy(buf, 16);

	return buf;
}

function createNTLMHash(password) {
	let md4sum = crypto.createHash('md4');
	md4sum.update(new Buffer(password, 'ucs2'));
	return md4sum.digest();
}

function createNTLMv2Hash(ntlmhash, username, authTargetName) {
	let hmac = crypto.createHmac('md5', ntlmhash);
	hmac.update(new Buffer(username.toUpperCase() + authTargetName, 'ucs2'));
	return hmac.digest();
}

function createLMv2Response(type2message, username, ntlmhash, nonce, targetName) {
	let buf = new Buffer(24),
		ntlm2hash = createNTLMv2Hash(ntlmhash, username, targetName),
		hmac = crypto.createHmac('md5', ntlm2hash);

	//server challenge
	type2message.challenge.copy(buf, 8);

	//client nonce
	buf.write(nonce || createPseudoRandomValue(16), 16, 'hex');

	//create hash
	hmac.update(buf.slice(8));
	let hashedBuffer = hmac.digest();

	hashedBuffer.copy(buf);

	return buf;
}

function createNTLMv2Response(type2message, username, ntlmhash, nonce, targetName) {
	let buf = new Buffer(48 + type2message.targetInfo.buffer.length),
		ntlm2hash = createNTLMv2Hash(ntlmhash, username, targetName),
		hmac = crypto.createHmac('md5', ntlm2hash);

	//the first 8 bytes are spare to store the hashed value before the blob

	//server challenge
	type2message.challenge.copy(buf, 8);

	//blob signature
	buf.writeUInt32BE(0x01010000, 16);

	//reserved
	buf.writeUInt32LE(0, 20);

	//timestamp
	//TODO: we are loosing precision here since js is not able to handle those large integers
	// maybe think about a different solution here
	// 11644473600000 = diff between 1970 and 1601
	let timestamp = ((Date.now() + 11644473600000) * 10000).toString(16);
	let timestampLow = Number('0x' + timestamp.substring(Math.max(0, timestamp.length - 8)));
	let timestampHigh = Number('0x' + timestamp.substring(0, Math.max(0, timestamp.length - 8)));

	buf.writeUInt32LE(timestampLow, 24, false);
	buf.writeUInt32LE(timestampHigh, 28, false);

	//random client nonce
	buf.write(nonce || createPseudoRandomValue(16), 32, 'hex');

	//zero
	buf.writeUInt32LE(0, 40);

	//complete target information block from type 2 message
	type2message.targetInfo.buffer.copy(buf, 44);

	//zero
	buf.writeUInt32LE(0, 44 + type2message.targetInfo.buffer.length);

	hmac.update(buf.slice(8));
	let hashedBuffer = hmac.digest();

	hashedBuffer.copy(buf);

	return buf;
}

function createPseudoRandomValue(length) {
	let str = '';
	while (str.length < length) {
		str += Math.floor(Math.random() * 16).toString(16);
	}
	return str;
}

module.exports = {
	createLMHash,
	createNTLMHash,
	createLMResponse,
	createNTLMResponse,
	createLMv2Response,
	createNTLMv2Response,
	createPseudoRandomValue
};

lib/ntlm.js
'use strict';

const os = require('os'),
	flags = require('./flags'),
	hash = require('./hash');

const NTLMSIGNATURE = "NTLMSSP\0";

function createType1Message(workstation, target) {
	let dataPos = 32,
		pos = 0,
		buf = new Buffer(1024);

	workstation = workstation === undefined ? os.hostname() : workstation;
	target = target === undefined ? '' : target;

	//signature
	buf.write(NTLMSIGNATURE, pos, NTLMSIGNATURE.length, 'ascii');
	pos += NTLMSIGNATURE.length;

	//message type
	buf.writeUInt32LE(1, pos);
	pos += 4;

	//flags
	buf.writeUInt32LE(flags.NTLMFLAG_NEGOTIATE_UNICODE |
		                flags.NTLMFLAG_NEGOTIATE_OEM |
						flags.NTLMFLAG_REQUEST_TARGET |
						flags.NTLMFLAG_NEGOTIATE_NTLM_KEY |
						flags.NTLMFLAG_NEGOTIATE_NTLM2_KEY |
						flags.NTLMFLAG_NEGOTIATE_ALWAYS_SIGN, pos);
	pos += 4;

	//domain security buffer
	buf.writeUInt16LE(target.length, pos);
	pos += 2;
	buf.writeUInt16LE(target.length, pos);
	pos += 2;
	buf.writeUInt32LE(target.length === 0 ? 0 : dataPos, pos);
	pos += 4;

	if (target.length > 0) {
		dataPos += buf.write(target, dataPos, 'ascii');
	}

	//workstation security buffer
	buf.writeUInt16LE(workstation.length, pos);
	pos += 2;
	buf.writeUInt16LE(workstation.length, pos);
	pos += 2;
	buf.writeUInt32LE(workstation.length === 0 ? 0 : dataPos, pos);
	pos += 4;

	if (workstation.length > 0) {
		dataPos += buf.write(workstation, dataPos, 'ascii');
	}

	return 'NTLM ' + buf.toString('base64', 0, dataPos);
}

function decodeType2Message(str) {
	if (str === undefined) {
		throw new Error('Invalid argument');
	}

	//convenience
	if (Object.prototype.toString.call(str) !== '[object String]') {
		if (str.hasOwnProperty('headers') && str.headers.hasOwnProperty('www-authenticate')) {
			str = str.headers['www-authenticate'];
		} else {
			throw new Error('Invalid argument');
		}
	}

	let ntlmMatch = /^NTLM ([^,\s]+)/.exec(str);

	if (ntlmMatch) {
		str = ntlmMatch[1];
	}

	let buf = new Buffer(str, 'base64'),
		obj = {};

	//check signature
	if (buf.toString('ascii', 0, NTLMSIGNATURE.length) !== NTLMSIGNATURE) {
		throw new Error('Invalid message signature: ' + str);
	}

	//check message type
	if (buf.readUInt32LE(NTLMSIGNATURE.length) !== 2) {
		throw new Error('Invalid message type (no type 2)');
	}

	//read flags
	obj.flags = buf.readUInt32LE(20);

	obj.encoding = (obj.flags & flags.NTLMFLAG_NEGOTIATE_OEM) ? 'ascii' : 'ucs2';

	obj.version = (obj.flags & flags.NTLMFLAG_NEGOTIATE_NTLM2_KEY) ? 2 : 1;

	obj.challenge = buf.slice(24, 32);

	//read target name
	obj.targetName = (function(){
		let length = buf.readUInt16LE(12);
		//skipping allocated space
		let offset = buf.readUInt32LE(16);

		if (length === 0) {
			return '';
		}

		if ((offset + length) > buf.length || offset < 32) {
			throw new Error('Bad type 2 message');
		}

		return buf.toString(obj.encoding, offset, offset + length);
	})();

	//read target info
	if (obj.flags & flags.NTLMFLAG_NEGOTIATE_TARGET_INFO) {
		obj.targetInfo = (function(){
			let info = {};

			let length = buf.readUInt16LE(40);
			//skipping allocated space
			let offset = buf.readUInt32LE(44);

			let targetInfoBuffer = new Buffer(length);
			buf.copy(targetInfoBuffer, 0, offset, offset + length);

			if (length === 0) {
				return info;
			}

			if ((offset + length) > buf.length || offset < 32) {
				throw new Error('Bad type 2 message');
			}

			let pos = offset;

			while (pos < (offset + length)) {
				let blockType = buf.readUInt16LE(pos);
				pos += 2;
				let blockLength = buf.readUInt16LE(pos);
				pos += 2;

				if (blockType === 0) {
					//reached the terminator subblock
					break;
				}

				let blockTypeStr;

				switch (blockType) {
					case 1:
						blockTypeStr = 'SERVER';
						break;
					case 2:
						blockTypeStr = 'DOMAIN';
						break;
					case 3:
						blockTypeStr = 'FQDN';
						break;
					case 4:
						blockTypeStr = 'DNS';
						break;
					case 5:
						blockTypeStr = 'PARENT_DNS';
						break;
					default:
						blockTypeStr = '';
						break;
				}

				if (blockTypeStr) {
					info[blockTypeStr] = buf.toString('ucs2', pos, pos + blockLength);
				}

				pos += blockLength;
			}

			return {
				parsed: info,
				buffer: targetInfoBuffer
			};
		})();
	}

	return obj;
}

function createType3Message(type2Message, username, password, workstation, target) {
	let dataPos = 52,
		buf = new Buffer(1024);

	if (workstation === undefined) {
		workstation = os.hostname();
	}

	if (target === undefined) {
		target = type2Message.targetName;
	}

	//signature
	buf.write(NTLMSIGNATURE, 0, NTLMSIGNATURE.length, 'ascii');

	//message type
	buf.writeUInt32LE(3, 8);

	if (type2Message.version === 2) {
		dataPos = 64;

		let ntlmHash = hash.createNTLMHash(password),
			nonce = hash.createPseudoRandomValue(16),
			lmv2 = hash.createLMv2Response(type2Message, username, ntlmHash, nonce, target),
			ntlmv2 = hash.createNTLMv2Response(type2Message, username, ntlmHash, nonce, target);

		//lmv2 security buffer
		buf.writeUInt16LE(lmv2.length, 12);
		buf.writeUInt16LE(lmv2.length, 14);
		buf.writeUInt32LE(dataPos, 16);

		lmv2.copy(buf, dataPos);
		dataPos += lmv2.length;

		//ntlmv2 security buffer
		buf.writeUInt16LE(ntlmv2.length, 20);
		buf.writeUInt16LE(ntlmv2.length, 22);
		buf.writeUInt32LE(dataPos, 24);

		ntlmv2.copy(buf, dataPos);
		dataPos += ntlmv2.length;
	} else {
		let lmHash = hash.createLMHash(password),
			ntlmHash = hash.createNTLMHash(password),
			lm = hash.createLMResponse(type2Message.challenge, lmHash),
			ntlm = hash.createNTLMResponse(type2Message.challenge, ntlmHash);

		//lm security buffer
		buf.writeUInt16LE(lm.length, 12);
		buf.writeUInt16LE(lm.length, 14);
		buf.writeUInt32LE(dataPos, 16);

		lm.copy(buf, dataPos);
		dataPos += lm.length;

		//ntlm security buffer
		buf.writeUInt16LE(ntlm.length, 20);
		buf.writeUInt16LE(ntlm.length, 22);
		buf.writeUInt32LE(dataPos, 24);

		ntlm.copy(buf, dataPos);
		dataPos += ntlm.length;
	}

	//target name security buffer
	buf.writeUInt16LE(type2Message.encoding === 'ascii' ? target.length : target.length * 2, 28);
	buf.writeUInt16LE(type2Message.encoding === 'ascii' ? target.length : target.length * 2, 30);
	buf.writeUInt32LE(dataPos, 32);

	dataPos += buf.write(target, dataPos, type2Message.encoding);

	//user name security buffer
	buf.writeUInt16LE(type2Message.encoding === 'ascii' ? username.length : username.length * 2, 36);
	buf.writeUInt16LE(type2Message.encoding === 'ascii' ? username.length : username.length * 2, 38);
	buf.writeUInt32LE(dataPos, 40);

	dataPos += buf.write(username, dataPos, type2Message.encoding);

	//workstation name security buffer
	buf.writeUInt16LE(type2Message.encoding === 'ascii' ? workstation.length : workstation.length * 2, 44);
	buf.writeUInt16LE(type2Message.encoding === 'ascii' ? workstation.length : workstation.length * 2, 46);
	buf.writeUInt32LE(dataPos, 48);

	dataPos += buf.write(workstation, dataPos, type2Message.encoding);

	if (type2Message.version === 2) {
		//session key security buffer
		buf.writeUInt16LE(0, 52);
		buf.writeUInt16LE(0, 54);
		buf.writeUInt32LE(0, 56);

		//flags
		buf.writeUInt32LE(type2Message.flags, 60);
	}

	return 'NTLM ' + buf.toString('base64', 0, dataPos);
}

module.exports = {
	createType1Message,
	decodeType2Message,
	createType3Message
};

lib/index.js
// Copyright (c) 2015 Nico Haller


// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:


// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.


// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.


'use strict';

const ntlmRequest = require('./ntlmrequest'),
	ntlm = require('./ntlm');


module.exports = {
	request: ntlmRequest,
	createType1Message: ntlm.createType1Message,
	decodeType2Message: ntlm.decodeType2Message,
	createType3Message: ntlm.createType3Message
};

lib/authenticate.js
'use strict';

const request = require('request'),
	extend = require('extend'),
	ntlm = require('./ntlm');


function authenticate(opts) {
	if (opts === undefined
		|| !'uri' in opts
		|| !'method' in opts
		|| !'username' in opts
		|| !'password' in opts) {
		throw new Error('Required options missing');
	}

	let ntlmState = {
			status: 0,
			data: null
		},
		ntlmRequest = request.defaults({
			jar: true,
			forever: true,
			headers: {
				'User-Agent': 'node.js',
				'Accept': '*/*',
				'Connection': 'keep-alive'
			}
		});

	function createRequestOptions(authHeader) {
		let obj = {
			uri: opts.uri,
			method: opts.method
		};

		if (opts.request) {
			extend(true, obj, opts.request, {
				headers: {
					'Authorization': authHeader
				}
			});
		}

		return obj;
	}

	function sendRequest(resolve, reject, err, res) {
		let authHeader;

		if (ntlmState.status === 0) {
			authHeader = ntlm.createType1Message();
		} else if (ntlmState.status === 1) {
			if (err || !res) {
				return reject(err || new Error('No response received'));
			}

			if (res.statusCode >= 400 && res.statusCode <= 499) {
				try {
					ntlmState.data = ntlm.decodeType2Message(res);
				} catch(e) {
					return reject(new Error('The server didnt respond properly'));
				}

				authHeader = ntlm.createType3Message(ntlmState.data, opts.username, opts.password, opts.workstation, opts.target);
			} else {
				return reject(new Error('The server apparently has no authentication enabled'));
			}
		} else if (ntlmState.status === 2) {
			if (err || !res || (res.statusCode < 200 && res.statusCode > 299)) {
				return reject(err || new Error('Authentication failed'));
			}

			return resolve(ntlmRequest);
		}

		ntlmState.status++;

		ntlmRequest(createRequestOptions(authHeader), sendRequest.bind(this, resolve, reject));
	}

	return new Promise(sendRequest);
}

module.exports = authenticate;

lib/ntlmrequest.js
'use strict';

const request = require('request'),
	extend = require('extend'),
	ntlm = require('./ntlm');


function ntlmRequest(opts) {
	if (opts === undefined
		|| !'uri' in opts
		|| !'method' in opts
		|| !'username' in opts
		|| !'password' in opts) {
		throw new Error('Required options missing');
	};

    opts.request = opts.request || {};

    // Handle case of username === 'SomeDomain\SomeUser'
    if(opts.username.includes('\\')) {
    	const splitUsername = opts.username.split('\\');
    	if (splitUsername.length !== 2) {
    		throw new Error('Unable to parse username');
		}
    	if(opts.target  && opts.target.toLocaleUpperCase() !== splitUsername[0].toLocaleUpperCase()) {
			throw new Error('Domain/target definition conflicts between vaule in username and value in target field.');
		}
		opts.target = splitUsername[0];
    	opts.username = splitUsername[1];
	}

	let ntlmState = {
			status: 0,
			data: null
		},
		ntlmRequestObj = request.defaults({
			jar: true,
			forever: true,
			headers: {
				'User-Agent': 'node.js',
				'Connection': 'keep-alive'
			}
		});

	function createRequestOptions(authHeader) {
		let obj = {
			uri: opts.uri,
			method: opts.method
		};

		extend(true, obj, opts.request, {
			headers: {
				'Authorization': authHeader
			}
		});

		return obj;
	}

	function sendRequest(resolve, reject, err, res, body) {
		let authHeader;

		if (ntlmState.status === 0) {
			authHeader = undefined;
		} else if (ntlmState.status === 1) {
			if (err || !res) {
				return reject(err || new Error('No response received'));
			}

			if (res.statusCode === 401) {
				let supportedAuthSchemes = [];

				for (let i = 0; i < res.rawHeaders.length; i = i + 2) {
					let headerName = res.rawHeaders[i].toLowerCase(),
						headerValue = res.rawHeaders[i + 1];

					if (headerName === 'www-authenticate') {
						headerValue.split(',').forEach((authMethod) => {
							authMethod = authMethod.trim();

							if (!/^[^\s=]+=.+/.test(authMethod)) {
								supportedAuthSchemes.push(authMethod.split(/\s/)[0].toLowerCase());
							}
						});
					}
				}

				if (supportedAuthSchemes.indexOf('ntlm') !== -1) {
					authHeader = ntlm.createType1Message();
				} else if (supportedAuthSchemes.indexOf('basic') !== -1) {
					authHeader = 'Basic ' + new Buffer(opts.username + ':' + opts.password).toString('base64');
					ntlmState.status = 2;
				} else {
					return reject(new Error('Could not negotiate on an authentication scheme'));
				}
			} else {
				return resolve({
					request: ntlmRequestObj,
					response: res,
					body: body
				});
			}
		} else if (ntlmState.status === 2) {
			if (err || !res) {
				return reject(err || new Error('No response received'));
			}

			if (res.statusCode >= 400 && res.statusCode <= 499) {
				try {
					ntlmState.data = ntlm.decodeType2Message(res);
				} catch(e) {
					return reject(new Error('The server didnt respond properly: ' + (e.stack || e.toString())));
				}

				authHeader = ntlm.createType3Message(ntlmState.data, opts.username, opts.password, opts.workstation, opts.target);
			} else {
				return resolve({
					request: ntlmRequestObj,
					response: res,
					body: body
				});
			}
		} else if (ntlmState.status === 3) {
            if (err || !res) {
                return reject(err || new Error('No response received'));
            }

			return resolve({
				request: ntlmRequestObj,
				response: res,
				body: body
			});
		}

		ntlmState.status++;

		ntlmRequestObj(createRequestOptions(authHeader), sendRequest.bind(this, resolve, reject));
	}

	return new Promise(sendRequest);
}

module.exports = ntlmRequest;
